Reworking UI of all menus

agzuniverse opened the pull request:
Reworked the UI of menus to achieve a more modern design consistent with Material Design guidelines.

Screenshot of settings screen before:
![settingscard_before](https://user-images.githubusercontent.com/22353313/38702990-b03d25be-3ebf-11e8-9577-96f2b7f28721.png)

Screenshot of settings screen after current changes:
![settingscard](https://user-images.githubusercontent.com/22353313/38772574-8ac9ce4c-4057-11e8-8f72-874b14494ebb.png)


In night mode:
![settingscard_night](https://user-images.githubusercontent.com/22353313/38772577-98ea936c-4057-11e8-85f1-09fc1bffa93e.png)


The `OptionDivider` component has been converted to a class based component in order to access `styles` via context like other common components.

Commit 8123dfc adds a padding around the version number in `DiagnosticsScreen` to make it look less crowded.
![vernumpadding](https://user-images.githubusercontent.com/22353313/38832104-25170f36-41df-11e8-88ee-3db5cad1c148.png)

gnprice commented:
Thanks @agzuniverse! I like this direction. And I appreciate the WIP pull request to get quick initial feedback.

Let's take the version where there's just one divider like in your last screenshot, but it goes the full width like in your second screenshot. One model that is good to default to is the Material Design style; take a look in particular at the [settings](https://material.io/guidelines/patterns/settings.html) doc on using dividers selectively to organize and making them full-width.

Also look at the [dividers](https://material.io/guidelines/components/dividers.html) doc for details on their appearance. These dividers look too visually strong for me in the screenshots, so I think a gray would be better; let's use the exact gray prescribed in Material Design, as that will be better than we're likely to pick ourselves.

I think if you make those tweaks, and then
* look at other settings screens that use these components, and add/remove OptionDividers there as needed, and
* add a screenshot showing how it looks in dark mode,

we'll be ready to start merging this change!

gnprice commented:
Oh, and of course fix the lint errors. :-)

gnprice commented:
Also, I just went ahead and merged the first of your commits :tada: , the small cleanup in Label. I'm always glad to merge parts of a PR early, as soon as they're ready -- it gives progress and it helps simplify code review for what's left.

agzuniverse commented:
Okay, I have made the `OptionDivider` a full width thin grey line in accordance with Material Design guidelines. I have updated the PR with a screenshot of the night mode as well. If this is all right, I can proceed to rework the other menus.

gnprice commented:
Thanks @agzuniverse ! This looks great.

I fired up this version in the emulator and took a quick look at the other menu screens (or precisely, the other screens that use OptionDivider), even without fixing up where they put the dividers. Some of them don't look great, but not really worse than the status quo -- so I'm going to go ahead and merge this. Please do go ahead and rework the other menus too. :-)

A few comments -- I'll fix these up before I merge, but for you to consider in the future:
* The last commit does a few other things beyond what the commit message says -- it doesn't just make it change color with theme, but also makes it a 1px line instead of a 16px space. That's sort of all part of the same conceptual change (making it a line instead of a space), so it makes sense to put it all in one commit -- but the commit message needs to clearly describe what it's actually doing.
* The last commit also changes the padding inside the items from 8px to 16px. That's easy to miss when reading the diff, so it definitely needs to be called out in the commit message -- and in fact it seems like a different change, so it should be a separate commit.
* Also I think I'd rather keep the option padding at 8px. :-) At 16px they feel too wide; and see e.g. the Material Design [guidelines on list items](https://material.io/guidelines/components/lists.html#lists-specs). One reason it's important to separate out different changes is so we can think about each of them like this in review.

agzuniverse commented:
Thanks for the thorough review and feedback @gnprice, I will ensure I am more verbose in my commit messages from now.

I have added 2 new commits, one reworking `OptionDivider` in every screen that uses that component right now, to divide groups of items and not every item.

The second commit fixes a small issue with `StreamScreen`. It used the padding prop of `Screen` to apply padding to everything inside, which meant the divider line could not extend the full width of the screen. I removed the usage of `padding` prop and wrapped the components that required padding in a View with the appropriate style applied instead.

borisyankov commented on src/streams/StreamScreen.js:
```diff
@@ -59,29 +60,31 @@ class StreamScreen extends PureComponent<Props> {
-        <StreamCard stream={stream} subscription={subscription} />
-        <OptionDivider />
+      <Screen title="Stream">
+        <View style={styles.padding}>
```
Why don't you apply this style to the `View` inside the `StreamCard`?

agzuniverse replied:
The approaches I could have taken are:

* Apply the padding to the `View` inside `StreamCard`, and apply it in `StreamScreen` for the buttons.

* Apply padding to both `StreamCard` and the buttons inside `StreamScreen` .

I chose the second approach as it felt cleaner and there is no 'mystery padding' to the `StreamCard` component. (None of the other components similar to `StreamCard` have padding presently.)

gnprice replied:
I think it actually makes sense for a component like `StreamCard` to contain its own padding. It really is part of the layout of that information (the icon, the stream name, the description) that it needs some appropriate padding around it.

If there are similar components that are relying on their environment to add padding, we might change those later. This one only has a single use site, so we don't need to worry about different usages of it having mismatched expectations.

agzuniverse replied:
Okay, I've made these changes. Now the `View` inside `StreamCard` contains padding, and it is no longer wrapped in a `View` just for padding in `StreamScreen` .

borisyankov commented on src/streams/StreamScreen.js:
```diff
@@ -59,29 +60,31 @@ class StreamScreen extends PureComponent<Props> {
-        <ZulipButton style={styles.marginTop} text="Topics" onPress={this.handleTopics} />
-        <ZulipButton style={styles.marginTop} text="Edit" onPress={this.handleEdit} />
+        <OptionDivider />
+        <View style={styles.screenPadding}>
```
Why do you need this? Adding padding here will misalign the buttons from both sides, as they already have padding.

agzuniverse replied:
But the reason I removed padding from `Screen` in the first place is so that the `OptionDivider` can span the full width of the screen (without having something weird like negative margins). If it did not it would look a bit odd.

borisyankov commented on src/styles/theme.js:
```diff
@@ -210,6 +210,9 @@ export default ({ color, backgroundColor, borderColor, cardColor, dividerColor }
     backgroundColor: cardColor,
     margin: 4,
   },
+  screenPadding: {
```
Why did you decide to name the style this way?

agzuniverse replied:
I named it `screenPadding` as it is padding applied to elements in a screen component, and can be used in the future for any screens that have components that already don't have padding ( Components such as `OptionRow` do, while `ZulipButton` does not)
However if this is too ambiguous, please do suggest a better name and I shall update the commit.

gnprice replied:
I think I prefer the direction of making components contain the padding that's essential to their layout -- like the padding that's already inside `OptionRow`. Then I think this `screenPadding` style item isn't necessary.

gnprice replied:
OK, we still have it for the buttons. I don't like this name either, though.

Following the naming scheme of lots of margin-related styles in this file, let's call this just `padding`. (I don't love that naming scheme, but changing it is for beyond this PR.)

There is an existing `padding` style, which by that naming scheme should be `smallPadding`. Happily, it looks like nothing is using that, so we can just remove it.

agzuniverse commented on src/streams/StreamScreen.js:
```diff
@@ -59,29 +60,31 @@ class StreamScreen extends PureComponent<Props> {
         />
-        <ZulipButton style={styles.marginTop} text="Topics" onPress={this.handleTopics} />
-        <ZulipButton style={styles.marginTop} text="Edit" onPress={this.handleEdit} />
+        <OptionDivider />
```
The idea of changing the style of `OptionDivider` was that it would be used to divide groups of settings, so using it to divide the options and the buttons seemed correct to me.

But if you are absolutely sure it is not required there, I will go ahead and remove it.

borisyankov replied:
Separately from that, imagine we wanted to add the border or any other effect.
That logic would be better extracted as a component.

`OptionGroup` can be just: 
```js
<View style={styles.optionGroup}>
```
And then you reuse it everywhere else that is needed.

gnprice commented:
Thanks @agzuniverse !

The changes in the three screens under settings look good, except I think we can leave out the divider at the bottom when there's nothing else below to divide things from.

One small followup in DiagnosticsScreen: since your previous changes, the version number at the top looks rather crowded. Would you give that some padding? That can be a separate commit.

I'll look closer at the StreamScreen commit later today -- sorry, short of time.

gnprice commented:
In StreamScreen:
* I agree that because of the dividers, it doesn't make sense for the `Screen` component to provide the left/right padding here.
* Note that `padding` means padding on all sides -- so if we put `padding` on a component in the middle of the screen, that adds padding above and below it too, which separates it from neighboring components.

  Here, that causes too much space between the divider and the first button:
![image](https://user-images.githubusercontent.com/28173/38828937-67d1e12a-416c-11e8-97e6-ef5b72244796.png)

Let's fix that extra space before we merge this change. I think the culprit is that 16px of `marginTop` that's on each of these buttons.

agzuniverse commented:
I've added padding around the version number in `DiagnosticsScreen` in a new commit.

Removing `marginTop` from the first button fixes the extra spacing on that screen. But the style has to be kept on the bottom button to avoid them being squashed together. This has no drawbacks other than a lack of symmetry, but if that's a problem both the buttons can be separately wrapped in a view that has just `paddingLeft` and `paddingRight` . (But that would add an extra view and requires defining a new style in `theme.js` .)

gnprice commented on src/streams/StreamScreen.js:
```diff
@@ -59,29 +60,29 @@ class StreamScreen extends PureComponent<Props> {
+        <OptionDivider />
+        <View style={styles.screenPadding}>
+          <ZulipButton text="Topics" onPress={this.handleTopics} />
+          <ZulipButton style={styles.marginTop} text="Edit" onPress={this.handleEdit} />
```
Hmm, I'm still not a fan of the `marginTop` on this button -- it feels like a pretty ad-hoc way of handling the layout. But that's how lots of other ZulipButtons around the app work, so (unlike with `StreamCard`) we can't just push the padding inside the component without fixing up lots of other use sites to match. So fixing that is beyond the scope of this PR.

agzuniverse commented:
Modified the commit. I verified that the existing style `padding` was not being used by anything and used that to apply padding to the buttons in `StreamScreen`, and removed the style `screenPadding` .

gnprice commented:
Looks great -- merged! Thanks @agzuniverse for all your work on this.

The only changes I made were in the commit message: some small formatting tweaks (notably, wrapping the paragraph to under 70 columns), and I added an explanation of the change to the `padding` style. The fact that the old style was unused is key for this change being correct, so it's good to mention it in the commit message.
